flax/core/lift.py

```diff
@@ -215,84 +216,48 @@ def repack(inner_scope_tree):
 id_fn = lambda x: x
 
 
-def transform(
+def map_variables(
     fn: Callable[..., Any],
-    target: CollectionFilter,
-    trans_in_fn: Callable[..., Any] = id_fn,
-    trans_out_fn: Callable[..., Any] = id_fn,
+    mapped_collections: CollectionFilter,
+    map_in_fn: Callable[..., Any] = id_fn,
+    map_out_fn: Callable[..., Any] = id_fn,
     init: bool = False, mutable: bool = False,
     rngs: PRNGSequenceFilter = True, variables: CollectionFilter = True):
-  """Locally transform Variables inside a scope.
+  """Locally map Variables inside a scope.
 
   Args:
     fn: the function to be transformed.
     target: the collection(s) to be transformed.
-    trans_in_fn: creates a view of the target variables.
-    trans_out_fn: transforms the updated variables in the view after mutation.
+    map_in_fn: creates a view of the target variables.
+    map_out_fn: transforms the updated variables in the view after mutation.
     init: If True, variables are initialized before transformation.
     rngs: PRNGSequences added to the transformed scope (default: all).
     variables: Additional Variable collections added to the transformed scope.
       Besides those specified by `target` (default: all).
   """
-  def wrapper(scope_fn, repack, variable_groups, rng_groups, treedef, *args):
+  def wrapper(scope_fn, repack, variable_groups, rng_groups, *args, **kwargs):
     target, variables = variable_groups
     if init:
-      scope = scope_fn((target, variables), rng_groups)
-      fn(scope, *args)
-      target, _ = repack(scope)
-      target_tree = trans_out_fn(treedef.unflatten(target))
-      target = treedef.flatten_up_to(target_tree)
-    target_tree = treedef.unflatten(map(unfreeze, target))
-    target_tree = trans_in_fn(target_tree)
-    target = treedef.flatten_up_to(target_tree)
+      scopes = scope_fn((target, variables), rng_groups)
+      has_mutable_cols = any(not is_filter_empty(scope.mutable) for scope in jax.tree_leaves(scopes))
+      if has_mutable_cols:
+        fn(scopes, *args, **kwargs)
+        target, _ = repack(scopes)
+        target = tuple(map_out_fn(x) for x in target)
+    target = tuple(map_in_fn(unfreeze(x)) for x in target)
     if not is_target_out:
       target = tuple(map(freeze, target))
-    scope = scope_fn((target, variables), rng_groups)
-    y = fn(scope, *args)
-    out_target, out_vars = repack(scope)
+    scopes = scope_fn((target, variables), rng_groups)
+    y = fn(scopes, *args, **kwargs)
+    out_target, out_vars = repack(scopes)
     if is_target_out:
-      out_target_tree = trans_out_fn(treedef.unflatten(out_target))
-      out_target = treedef.flatten_up_to(out_target_tree)
+      out_target = tuple(map_out_fn(x) for x in out_target)
     return y, (out_target, out_vars)
 
   is_target_out = mutable or init
```

Comment:
move `is_target_out` definition above its uses in conditionals ^ ^ ^

Reply:
is_target_out is used outside the wrapper as well to determine the out_variable_groups

Reply:
oh sure, I know it works in scope, I was just suggesting lexically to move it to the top before any uses of it.

Reply:
Fair enough